lib/layout/force-directed-layout.js

```diff
@@ -0,0 +1,202 @@
+		attractionX = Math.random()*10+1;
+		attractionY = Math.random()*10+1;
+	} else {
+		attractionX = xLength/(length*length);
+		attractionY = yLength/(length*length);
+	}
+	var len = attractionX * attractionX + attractionY * attractionY;
+	//ensures length is positive
+	if(len>0){
+		len = Math.sqrt(len);
+		attractionX = attractionX / len;
+		attractionY = attractionY / len;
+	}
+	//repulsion minus attraction
+	repulsionX = (repulsionX - attractionX);
+	repulsionY = repulsionY - attractionY;
+	var lim = 0.1;
+	
+	if(vertexTo.get('isPinned') ===false && vertexFrom.get('isPinned')===false){
+		vertexTo.set('position', {
+								'x': (vertexTo.get('position').x+repulsionX),
+								'y': (vertexTo.get('position').y+repulsionY)
+								});
+		vertexFrom.set('position', {
+								'x':vertexFrom.get('position').x-repulsionX,
+								'y':vertexFrom.get('position').y-repulsionY
+								});
+	}else if(vertexTo.get('isPinned') ===true){
+		vertexFrom.set('position', {
+								'x':vertexFrom.get('position').x-repulsionX-repulsionX, 
+								'y':vertexFrom.get('position').y-repulsionY-repulsionY
+								});
+	
+	}else if(vertexFrom.get('isPinned')===true){
+		vertexTo.set('position', {
+								'x': (vertexTo.get('position').x+repulsionX+repulsionX),
+								'y': (vertexTo.get('position').y+repulsionY+repulsionY)
+								});
+	}
+	if(repulsionX > lim || repulsionY > lim || (repulsionX < -lim) || (repulsionY < -lim)){
+			change = true;
+	}
+}
+/* not yet implemented
+function portPositioning(edge){
+	var vertexTo = mGraph.getVertexById(edge.get('to').get('ownerVertexId'));
+	var vertexFrom = mGraph.getVertexById(edge.get('from').get('ownerVertexId'));
+	var xLength = edgeXLength(edge);
+	var yLength = edgeYLength(edge);
+	
+
+
+}*/
+
+//calculates edge length
+function edgeLength(edge){
+	to=mGraph.getVertexById(edge.get('to').get('ownerVertexId'));
+	from=mGraph.getVertexById(edge.get('from').get('ownerVertexId'));
+	
+	if('null' !=to.get('position').x&&'null' !=to.get('position').y&&'null' !=from.get('position').x&& 'null' != from.get('position').y){
+		//pythag theorem 
+		var x = to.get('position').x - from.get('position').x;
+		var y = to.get('position').y - from.get('position').y;
+		
+		var c = x*x+y*y;
+		return Math.sqrt(c);
+	}else{
+	//@todo add error checking
+	//on fail return 0 length
+		return 0;
+	
+	}
+
+}
+
+function edgeXLength(edge){
+	to=mGraph.getVertexById(edge.get('to').get('ownerVertexId'));
+	from=mGraph.getVertexById(edge.get('from').get('ownerVertexId'));
+	
+	if('null' !=to.get('position').x&&'null' !=from.get('position').x){
+		var x = to.get('position').x - from.get('position').x;
+		return x;
+	}else{
+	//@todo add error checking
+	//on fail return 0 length
+		return 0;
+	
+	}
+
+}
+
+function edgeYLength(edge){
+	to=mGraph.getVertexById(edge.get('to').get('ownerVertexId'));
+	from=mGraph.getVertexById(edge.get('from').get('ownerVertexId'));
+	
+	if('null' !=to.get('position').y&&'null' !=from.get('position').y){
```

Comment:
![Yoda condition](http://i.imgur.com/F7P39Qv.jpg)

And why are you checking for the string 'null' anyways?